walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,79 @@
 
 package com.google.sps;
 
+import java.util.*; 
 import java.util.Collection;
+import com.google.sps.TimeRange;
+import com.google.common.collect.Iterables;
 
+/** 
+ * Determines a list of available time slots to schedule a meeting given
+ * already scheduled meetings. 
+ */
 public final class FindMeetingQuery {
+  
+  /** 
+   * Returns the open windows for which {@code request} can be scheduled in
+   * while taking account conflicting meetings in {@code events}.
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requiredAttendees = request.getAttendees();
+    Collection<TimeRange> availableTimes = assembleTime(events, requiredAttendees, request.getDuration());
+    return adjustAvailableTimes;
+  }
+
+  /**
+   * Returns the open windows for which {@code request} can be scheduled in,
+   * after making {@code events} TimeRanges unavailable in {@code availableTimes}.
+   */
+  public Collection<TimeRange> assembleTime(Collection<Event> events, Collection<String> requiredAttendees, long duration) {
+    Collection<TimeRange> availableTimes = new ArrayList<>();
+
+    // Initally, entire day is open to schedule a meeting in.
+    availableTimes.add(TimeRange.WHOLE_DAY);
+
+    // For every event that a requiredAttendee must attended, remove that TimeRange from availableTimes.
+    for(Event currEvent : events) {
+      if (!Collections.disjoint(currEvent.getAttendees(), requiredAttendees)) {
+        availableTimes = adjustAvailableTimes(availableTimes, currEvent.getWhen());
+      }
+    }
+
+    // Remove TimeRanges that are too short.
+    availableTimes.removeIf(item -> item.duration() < duration);
+
+    return availableTimes;
+  }
+
+  /**
+   * Returns {@code newTimes} which takes all TimeRanges in {@code availableTimes} and
```

Comment:
Here too, you should not reference internal variables. Please rephrase this comment.

Reply:
Fixed